crates/anstyle-roff/src/style_stream.rs

```diff
@@ -0,0 +1,229 @@
+        style = style.effects(effects);
+
+        Self {
+            text: category.text,
+            style,
+        }
+    }
+}
+
+fn create_effects(category: &CategorisedSlice) -> Effects {
+    Effects::new()
+        .set(Effects::ITALIC, category.italic.unwrap_or(false))
+        .set(Effects::BLINK, category.blink.unwrap_or(false))
+        .set(Effects::INVERT, category.reversed.unwrap_or(false))
+        .set(Effects::HIDDEN, category.hidden.unwrap_or(false))
+        .set(
+            Effects::STRIKETHROUGH,
+            category.strikethrough.unwrap_or(false),
+        )
+        .set(Effects::UNDERLINE, category.underline.unwrap_or(false))
+        .set(Effects::BOLD, is_bold(category.intensity))
+        .set(Effects::DIMMED, is_faint(category.intensity))
+}
+
+fn is_bold(intensity: Option<Intensity>) -> bool {
+    matches!(intensity, Some(Intensity::Bold))
+}
+
+fn is_faint(intensity: Option<Intensity>) -> bool {
+    matches!(intensity, Some(Intensity::Faint))
+}
+
+impl<'a> IntoIterator for StyledStream<'a> {
+    type Item = StyledStr<'a>;
+    type IntoIter = <Vec<StyledStr<'a>> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.stream.into_iter()
+    }
+}
+
+impl<'a> Deref for StyledStream<'a> {
+    type Target = [StyledStr<'a>];
+
+    fn deref(&self) -> &Self::Target {
+        &self.stream[..]
+    }
+}
+
+impl<'a> DerefMut for StyledStream<'a> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.stream[..]
+    }
+}
+
+fn cansi_to_anstyle_color(color: Option<Color>) -> Option<AColor> {
+    match color {
+        Some(Color::Black) => Some(AColor::Ansi(AnsiColor::Black)),
+        Some(Color::Red) => Some(AColor::Ansi(AnsiColor::Red)),
+        Some(Color::Green) => Some(AColor::Ansi(AnsiColor::Green)),
+        Some(Color::Yellow) => Some(AColor::Ansi(AnsiColor::Yellow)),
+        Some(Color::Blue) => Some(AColor::Ansi(AnsiColor::Blue)),
+        Some(Color::Magenta) => Some(AColor::Ansi(AnsiColor::Magenta)),
+        Some(Color::Cyan) => Some(AColor::Ansi(AnsiColor::Cyan)),
+        Some(Color::White) => Some(AColor::Ansi(AnsiColor::White)),
+        Some(Color::BrightBlack) => Some(AColor::Ansi(AnsiColor::BrightBlack)),
+        Some(Color::BrightRed) => Some(AColor::Ansi(AnsiColor::BrightRed)),
+        Some(Color::BrightGreen) => Some(AColor::Ansi(AnsiColor::BrightGreen)),
+        Some(Color::BrightYellow) => Some(AColor::Ansi(AnsiColor::BrightYellow)),
+        Some(Color::BrightBlue) => Some(AColor::Ansi(AnsiColor::BrightBlue)),
+        Some(Color::BrightMagenta) => Some(AColor::Ansi(AnsiColor::BrightMagenta)),
+        Some(Color::BrightCyan) => Some(AColor::Ansi(AnsiColor::BrightCyan)),
+        Some(Color::BrightWhite) => Some(AColor::Ansi(AnsiColor::BrightWhite)),
+        None => None,
+    }
+}
+
+impl<'text> StyledStream<'text> {
+    pub(crate) fn new(s: &'text str) -> Self {
+        let categorized = cansi::v3::categorise_text(s);
```

Comment:
My thinking was to isolate the use of cansi to one module in case we want to change the provider for categorizing text. Not using a whole other struct is feasable though. I wrote a func to just return an iterator of `StyledStr`.